Bolts/Common/BFTask.m

```diff
@@ -229,84 +236,114 @@ + (instancetype)taskFromExecutor:(BFExecutor *)executor withBlock:(nullable id (
-    @synchronized(self.lock) {
-        if (self.completed) {
-            return NO;
+    __block BOOL rval;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        if (_state != pending) {
+            rval = NO;
+            return;
         }
-        self.completed = YES;
+        _state = succeeded;
         _result = result;
+        rval = YES;
+    });
+    if (rval) {
         [self runContinuations];
-        return YES;
     }
+    return rval;
 }
 
 - (nullable NSError *)error {
-    @synchronized(self.lock) {
-        return _error;
-    }
+    __block NSError *error;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        error = _error;
+    });
+    return _error;
 }
 
 - (BOOL)trySetError:(NSError *)error {
-    @synchronized(self.lock) {
-        if (self.completed) {
-            return NO;
+    __block BOOL rval;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        if (_state != pending) {
+            rval = NO;
+            return;
         }
-        self.completed = YES;
-        self.faulted = YES;
+        _state = errored;
         _error = error;
+        rval = YES;
+    });
+    if (rval) {
         [self runContinuations];
-        return YES;
     }
+    return rval;
 }
 
 - (BOOL)isCancelled {
-    @synchronized(self.lock) {
-        return _cancelled;
-    }
+    __block BOOL isCancelled;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        isCancelled = _state == cancelled;
+    });
+    return isCancelled;
 }
 
 - (BOOL)isFaulted {
-    @synchronized(self.lock) {
-        return _faulted;
-    }
+    __block BOOL faulted;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        faulted = _state == errored;
+    });
+    return faulted;
 }
 
 - (BOOL)trySetCancelled {
-    @synchronized(self.lock) {
-        if (self.completed) {
-            return NO;
+    __block BOOL rval;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        if (_state != pending) {
+            rval = NO;
+            return;
         }
-        self.completed = YES;
-        self.cancelled = YES;
+        _state = cancelled;
+        rval = YES;
+    });
+    if (rval) {
         [self runContinuations];
-        return YES;
     }
+    return rval;
 }
 
 - (BOOL)isCompleted {
-    @synchronized(self.lock) {
-        return _completed;
-    }
+    __block BOOL completed;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        completed = _state != pending;
+    });
+    return completed;
 }
 
 - (void)runContinuations {
-    @synchronized(self.lock) {
+    __block NSArray* callbacks;
+    dispatch_barrier_sync(_synchronizationQueue, ^{
         [self.condition lock];
         [self.condition broadcast];
         [self.condition unlock];
-        for (void (^callback)(void) in self.callbacks) {
-            callback();
-        }
+        callbacks = [NSArray arrayWithArray:self.callbacks];
         [self.callbacks removeAllObjects];
+    });
+    for (void (^callback)() in callbacks) {
```

Comment:
There was a reason IIRC that we had the callbacks executing in the context of self.lock before. I believe it had to do with a race whereupon you could enqueue a new callback in-between `-removeAllObjects` and execution of the callbacks, and allow another continuation to execute out-of-order.

If I recall from the conversations we had at the time, we weren't comfortable changing those guarantees for people who were using Bolts already, whereas Bolts-Swift was able to break those guarantees as they didn't exist yet.

I'm OK with switching it to this version, I just wanted to make sure the context is documented here.

Reply:
OK, I recall I had issues with keeping the execution in the sync block, perhaps outside a barrier is better, LMK.

Reply:
@richardjrossiii this seems to lead to a deadlock in that test

https://github.com/parse-community/Parse-SDK-iOS-OSX/blob/7a820b75c6726808475af9ce705053ff0e8cbd11/Tests/Unit/SQLiteDatabaseTest.m#L494